client/components/domains/map-domain-step/index.jsx

```diff
@@ -20,16 +20,17 @@ var MapDomainStep = React.createClass( {
 
 	propTypes: {
 		products: React.PropTypes.object.isRequired,
+		initialQuery: React.PropTypes.string,
 		analyticsSection: React.PropTypes.string.isRequired
 	},
 
 	getInitialState: function() {
-		return { searchQuery: '' };
+		return { searchQuery: this.props.initialQuery };
```

Comment:
I really like the idea! Seems like the proper React way to handle this case.
Unfortunately, the way signup is written now, the state is saved [when the component is unmounted](https://github.com/Automattic/wp-calypso/blob/fix/autofill-domain-mapping/client/components/domains/map-domain-step/index.jsx#L47-L51) (and restored when needed later). This makes it necessary to keep the search query in state either way :/ We could still do it your way - but it would make it ambiguous where the query should be (in state or props) and would add additional complexity.